src/rules/noMagicNumbersRule.ts

```diff
@@ -0,0 +1,81 @@
+/**
+ * @license
+ * Copyright 2016 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as ts from "typescript";
+
+import * as Lint from "../index";
+
+export class Rule extends Lint.Rules.AbstractRule {
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "no-magic-numbers",
+        description: "Disallows the use constant number values outside of variable assignment. -1, 0 and 1 are allowed by default.",
+        rationale: Lint.Utils.dedent`
+            Magic numbers should be avoided as they often lack documentation, forcing
+            them to be stored in variables gives them implicit documentation.`,
+        optionsDescription: "A list of allowed number.",
+        options: {
+            type: "array",
+            items: {
+                type: "number",
+            },
+            minLength: 1,
+        },
+        optionExamples: ["true", "[true, 1, 2, 3]"],
+        type: "typescript",
+        typescriptOnly: false,
+    };
+    /* tslint:enable:object-literal-sort-keys */
+
+    public static FAILURE_STRING = "'magic numbers' are no allowed";
+
+    public static ALLOWED_NODES = [
+        ts.SyntaxKind.ExportAssignment,
+        ts.SyntaxKind.FirstAssignment,
+        ts.SyntaxKind.LastAssignment,
+        ts.SyntaxKind.PropertyAssignment,
+        ts.SyntaxKind.ShorthandPropertyAssignment,
+        ts.SyntaxKind.VariableDeclaration,
+        ts.SyntaxKind.VariableDeclarationList,
+        ts.SyntaxKind.EnumMember,
+        ts.SyntaxKind.PropertyDeclaration,
+    ];
+
+    public static DEFAULT_ALLOWED = [
+        0,
+        1,
+    ];
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithWalker(new NoMagicNumbersWalker(sourceFile, this.getOptions()));
+    }
+}
+
+class NoMagicNumbersWalker extends Lint.RuleWalker {
+    public visitNode(node: ts.Node) {
+        if (node.kind === ts.SyntaxKind.NumericLiteral &&
+            Rule.ALLOWED_NODES.indexOf(node.parent.kind) === -1) {
+            const options = this.getOptions();
+
+            const allowed: number[] = options.length > 0 ? options : Rule.DEFAULT_ALLOWED;
```

Comment:
read `options` and do this logic in a constructor so you don't have to do this for every node